### PR TITLE
[PVR] Rework component startup / shutdown

### PR DESCRIPTION
**Pull request by ksooo**
This fixes and improves a couple of things related to PVR startup / shutdown. Sorry, I was not able to split this up in separate commits as code changes evolved over time, in many iterations.

1) Fixes a startup race which happens with multiple PVR add-ons enabled, and at least one of the add-ons supports "async connect". 

If the add-on supporting async-connect notifices "connected" somewhere in the middle of pvr manager loading components, it could happen that PVR ended up with inconsistent data. Loading sequence is GetChannelGroups, GetChannels, for every group GetChannelGroupMembers. If for example, call to GetChannelGroups fails, call to GetChannels also fails because add-on supporting async-connect is not yet connected, then the add-ons gets connected, call to first and subsequent GetChannelGroupMembers will succeed, leading evtually to deletion of all channel data in the PVR data base. => Log entries ala "Removed stale channel...)

The race gets fixed by determining the ready-to-use add-ons and passing those add-ons to the respective methods which are changed to load data from the database for the given clients only and to update date from the given clients only. PVR manager main loop now periodically checks for newly connected clients and in case to load data for those new clients only.

2) Due to the beforementioned code changes, PVR restart no longer needs to get restarted when an add-on signals "connected".

3) Data update jobs are now only scheduled for the add-on requesting to update data only, not for all available PVR clients.

4) Some logging improvements

I runtime-tested the changes very carefully on macOS and Android - no regressions found.

@phunkyfish when you find some time for a code review...

### CODE REVIEW

**Review by lrusak — COMMENTED** [review]
Had a quick look through, can't really comment on the functionality of the code but it sounds like a nice improvement. It looks like there is a good opportunity to constify some methods and I pointed out a few opportunities for future changes.

Good work 👍

File: xbmc/pvr/PVRDatabase.cpp [hunk 1]
```diff
@@ -91,6 +91,26 @@ namespace
     ")";
 
   // clang-format on
+
+std::string GetClientIdsSQL(const std::vector<std::shared_ptr<CPVRClient>>& clients)
+{
+  std::string clientIds;
+  if (!clients.empty())
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.cpp]: Yep, will change.

File: xbmc/pvr/PVRDatabase.cpp [hunk 2]
```diff
@@ -91,6 +91,26 @@ namespace
...
+  {
+    clientIds += "(";
+    for (auto it = clients.cbegin(); it != clients.cend(); ++it)
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.cpp]: > I came up with this but it's not really an improvement

I do agree. :-) I will keep this as it it is.

File: xbmc/pvr/PVRDatabase.cpp [hunk 3]
```diff
@@ -646,22 +678,24 @@ std::vector<std::shared_ptr<CPVRChannelGroupMember>> CPVRDatabase::Get(
...
-                 "map_channelgroups_channels.iOrder, "
-                 "map_channelgroups_channels.iClientChannelNumber, "
-                 "map_channelgroups_channels.iClientSubChannelNumber, "
-                 "channels.iClientId, channels.iUniqueId, channels.bIsRadio "
-                 "FROM map_channelgroups_channels "
-                 "LEFT JOIN channels ON channels.idChannel = map_channelgroups_channels.idChannel "
-                 "WHERE map_channelgroups_channels.idGroup = %i ORDER BY "
-                 "map_channelgroups_channels.iChannelNumber",
-                 group.GroupID());
+  std::string strQuery =
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.cpp:681]: > Something to consider

Maybe in a followup-PR.

File: xbmc/pvr/PVRDatabase.h [hunk 4]
```diff
@@ -114,10 +114,13 @@ namespace PVR
...
      * @brief Get channels from the database.
      * @param bRadio Whether to fetch radio or TV channels.
+     * @param clients The PVR clients the channels should be loaded for. Leave empty for all clients.
      * @param results The container for the channels.
      * @return The number of channels loaded.
      */
-    int Get(bool bRadio, std::map<std::pair<int, int>, std::shared_ptr<CPVRChannel>>& results);
+    int Get(bool bRadio,
+            const std::vector<std::shared_ptr<CPVRClient>>& clients,
+            std::map<std::pair<int, int>, std::shared_ptr<CPVRChannel>>& results);
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h:123]: yep

File: xbmc/pvr/PVRDatabase.h [hunk 5]
```diff
@@ -175,9 +178,10 @@ namespace PVR
     /*!
      * @brief Get the list of providers from the database
      * @param results The providers to store the results in.
+     * @param clients The PVR clients the providers should be loaded for. Leave empty for all clients.
      * @return The amount of providers that were added.
      */
-    bool Get(CPVRProviders& results);
+    bool Get(CPVRProviders& results, const std::vector<std::shared_ptr<CPVRClient>>& clients);
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h:184]: yep

File: xbmc/pvr/PVRDatabase.h [hunk 6]
```diff
@@ -207,9 +211,11 @@ namespace PVR
     /*!
      * @brief Get the members of a channel group.
      * @param group The group to get the members for.
+     * @param clients The PVR clients the group members should be loaded for. Leave empty for all clients.
      * @return The group members.
      */
-    std::vector<std::shared_ptr<CPVRChannelGroupMember>> Get(const CPVRChannelGroup& group);
+    std::vector<std::shared_ptr<CPVRChannelGroupMember>> Get(
+        const CPVRChannelGroup& group, const std::vector<std::shared_ptr<CPVRClient>>& clients);
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h:218]: yeap

File: xbmc/pvr/PVRDatabase.h [hunk 7]
```diff
@@ -230,9 +236,11 @@ namespace PVR
     /*!
      * @brief Get the timers.
      * @param timers The container for the timers.
+     * @param clients The PVR clients the timers should be loaded for. Leave empty for all clients.
      * @return The timers.
      */
-    std::vector<std::shared_ptr<CPVRTimerInfoTag>> GetTimers(CPVRTimers& timers);
+    std::vector<std::shared_ptr<CPVRTimerInfoTag>> GetTimers(
+        CPVRTimers& timers, const std::vector<std::shared_ptr<CPVRClient>>& clients);
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h:243]: okay

File: xbmc/pvr/PVRManager.cpp [hunk 8]
```diff
@@ -659,47 +662,127 @@ void CPVRManager::OnWake()
...
+    progressHandler = nullptr; // no delete, instance is deleting itself
+  }
+}
+
+bool CPVRManager::UpdateComponents(std::vector<std::shared_ptr<CPVRClient>>& knownClients,
+                                   ManagerState stateToCheck,
+                                   CPVRGUIProgressHandler* progressHandler)
+{
+  // find clients which disappeared since last check and remove them from known clients
+  for (auto it = knownClients.begin(); it != knownClients.end();)
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRManager.cpp:697]: Will keep it as it is. :-)

File: xbmc/pvr/PVRManager.h [hunk 9]
```diff
@@ -381,12 +386,46 @@ namespace PVR
      */
     bool SetWakeupCommand();
 
+    enum ManagerState
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRManager.h:389]: Just moved existing code, but yes.

File: xbmc/pvr/addons/PVRClient.cpp [hunk 10]
```diff
@@ -171,11 +171,15 @@ ADDON_STATUS CPVRClient::Create(int iClientId)
...
-  bool bReadyToUse(false);
-  CLog::LogFC(LOGDEBUG, LOGPVR, "Creating PVR add-on instance '{}'", Name());
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Creating PVR add-on instance '{}'", ID());
+
+  bool bReadyToUse = false;
   if ((status = CreateInstance()) == ADDON_STATUS_OK)
     bReadyToUse = GetAddonProperties();
 
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Created PVR add-on instance '{}'. readytouse={} ", ID(),
+              bReadyToUse ? "true" : "false");
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/addons/PVRClient.cpp:181]: cool, thanks.

File: xbmc/pvr/PVRDatabase.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h]: Does not work without further changes. Method calls `CDatabase::ResultQuery` which is not `const`. Will not change this in this PR:

File: xbmc/pvr/PVRDatabase.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h]: Does not work without further changes. Method calls `CDatabase::ResultQuery` which is not `const`. Will not change this in this PR:

File: xbmc/pvr/PVRDatabase.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h]: Does not work without further changes. Method calls `CDatabase::ResultQuery` which is not `const`. Will not change this in this PR:

File: xbmc/pvr/PVRDatabase.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRDatabase.h]: Does not work without further changes. Method calls `CDatabase::ResultQuery` which is not `const`. Will not change this in this PR:

**Comment by ksooo** [comment]
jenkins build this please

**Review by lrusak — APPROVED** [review]
Thanks for going the extra mile with the enum class and the constification efforts :+1:

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by ksooo** [comment]
jenkins build this please

File: xbmc/pvr/PVRManager.h [hunk 11]
```diff
@@ -276,8 +276,9 @@ namespace PVR
 
     /*!
      * @brief Let the background thread update the recordings list.
+     * @param clientId The id of the PVR client to update.
      */
-    void TriggerRecordingsUpdate();
+    void TriggerRecordingsUpdate(int clientId = PVR_INVALID_CLIENT_ID);
```
Comment (phunkyfish) [inline xbmc/pvr/PVRManager.h:269]: Might be more clear to have two functions and skip the default value.

**Review by phunkyfish — APPROVED** [review]
Nothing blocking. Just something someone once told me about default values in functions :)

**Comment by ksooo** [comment]
> someone once told me about default values in functions :)

 Caught me. :-)

File: xbmc/pvr/PVRManager.h (hunk 11 above)
Comment (ksooo) [reply xbmc/pvr/PVRManager.h:269]: Fair.

**Comment by lrusak** [comment]
jenkins build this please

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
jenkins build this please

**Pull request merged**